silo/api.py

```diff
@@ -302,28 +300,59 @@ def data(self, request, id):
         # calling get_object applies the permission classes to this query
         self.get_object()
 
+        # get the mongo collection
+        collection = LabelValueStore._get_collection()
+
         draw = int(request.GET.get("draw", 1))
         offset = int(request.GET.get('start', -1))
         length = int(request.GET.get('length', 10))
 
         # filtering syntax is the mongodb syntax
-        query = request.GET.get('query',"{}")
-        filter_fields = json.loads(query)
-
-        recordsTotal = LabelValueStore.objects(silo_id=id, **filter_fields).count()
-
-        # workaround until the problem of javascript not increasing the value of length is fixed
+        query = request.GET.get('query', '{}')
+        group = request.GET.get('group', '{}')
+        sort = str(request.GET.get('sort', '{}'))
+        query_fields = json.loads(query)
+        group_fields = json.loads(group)
+        sort_fields = json.loads(sort)
+
+        # creating the aggregation pipeline
+        pipeline = [
+            {'$match': {'$and': [{'silo_id': int(id)}, query_fields]}},
+            {'$project': {
+                'create_date': 0,
+                'edit_date': 0,
+                'silo_id': 0,
+                'read_id': 0
+            }}
+        ]
+        if group_fields:
+            pipeline.append({'$group': group_fields})
+        if sort_fields:
+            pipeline.append({'$sort': sort_fields})
+
+        count_pipeline = pipeline + [{'$count': 'count'}]
+        cmd_cur = collection.aggregate(pipeline=count_pipeline)
+        result = list(cmd_cur)[0]
+        records_total = result['count']
+
+        # workaround until the problem of javascript not increasing
+        # the value of length is fixed
         if offset >= 0:
```

Comment:
If you know the `offset` before, finish appending to the pipeline before you start doing operations with the database

Reply:
- This `offset` is only used for the calculation of the length the if.
- `offset` action in MongoDB doesn't accept negative values as well